### PR TITLE
CATROID-876 Allow to edit bricks in Catblocks

### PR DESCRIPTION
**Pull request by bprattes**
https://jira.catrob.at/browse/CATROID-876
Now it is possible to edit bricks in Catblocks (add, remove, move, duplicate).
Therefore each brick has its own unique ID.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/DummyBrick.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    override fun clone(): Brick {
+        val clone = super.clone() as DummyBrick
+        clone.script = script.clone() as DummyScript
+        clone.script.scriptBrick = clone
+        return clone
+    }
+
+    override fun addActionToSequence(sprite: Sprite?, sequence: ScriptSequenceAction?) {
+        // do nothing
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/DummyBrick.kt:55]: delete comment.
If lint acts up add a ignore

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java [hunk 2]
```diff
@@ -197,4 +202,107 @@ protected String getPositionInformation() {
...
+				return true;
+			} else if (subStackIndex == 1 && cb.hasSecondaryList()) {
+				cb.getSecondaryNestedBricks().addAll(0, bricksToAdd);
+				return true;
+			}
+		}
+
+		int idx = 0;
+
+		for (Brick b : cb.getNestedBricks()) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java:280]: variable name b - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java [hunk 3]
```diff
@@ -197,4 +202,107 @@ protected String getPositionInformation() {
...
+
+	@Override
+	public boolean addBrickInNestedBrick(UUID parentBrickId, int subStackIndex, List<Brick> bricksToAdd) {
+		if (!(this instanceof CompositeBrick)) {
+			return false;
+		}
+
+		// if subStackIndex == -1, the first brick to add is no first sub stack item!
+
+		CompositeBrick cb = (CompositeBrick) this;
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java:266]: variable name

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java [hunk 4]
```diff
@@ -197,4 +202,107 @@ protected String getPositionInformation() {
...
+			if (subStackIndex == 0) {
+				cb.getNestedBricks().addAll(0, bricksToAdd);
+				return true;
+			} else if (subStackIndex == 1 && cb.hasSecondaryList()) {
+				cb.getSecondaryNestedBricks().addAll(0, bricksToAdd);
+				return true;
+			}
+		}
+
+		int idx = 0;
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java:278]: variable name. Is it a typo of id?

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java [hunk 5]
```diff
@@ -197,4 +202,107 @@ protected String getPositionInformation() {
...
+				}
+			}
+
+			if (brickIds.size() == foundBricks.size()) {
+				break;
+			}
+		}
+
+		if (foundBricks.size() == 0 && cb.hasSecondaryList()) {
+			for (Brick b : cb.getSecondaryNestedBricks()) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java:236]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java [hunk 6]
```diff
@@ -197,4 +202,107 @@ protected String getPositionInformation() {
...
+	@Override
+	public List<Brick> findBricksInNestedBricks(List<UUID> brickIds) {
+		if (!(this instanceof CompositeBrick)) {
+			return null;
+		}
+
+		List<Brick> foundBricks = new ArrayList<>();
+		CompositeBrick cb = (CompositeBrick) this;
+
+		for (Brick b : cb.getNestedBricks()) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java:220]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java [hunk 7]
```diff
@@ -197,4 +202,107 @@ protected String getPositionInformation() {
...
+	}
+
+	@Override
+	public List<Brick> findBricksInNestedBricks(List<UUID> brickIds) {
+		if (!(this instanceof CompositeBrick)) {
+			return null;
+		}
+
+		List<Brick> foundBricks = new ArrayList<>();
+		CompositeBrick cb = (CompositeBrick) this;
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/bricks/BrickBaseType.java:218]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/Sprite.java [hunk 8]
```diff
@@ -573,4 +573,22 @@ public void releaseAllPointers() {
 		}
 		usedTouchPointer.clear();
 	}
+
+	public List<UUID> removeAllEmptyDummyScripts() {
+		List<UUID> idsToRemove = new ArrayList<>();
+		List<Script> scriptsToRemove = new ArrayList<>();
+
+		for (Script s : scriptList) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/Sprite.java:581]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/Script.java [hunk 9]
```diff
@@ -219,4 +219,73 @@ public void deselectElements(List<UserData<?>> elements) {
...
+					break;
+				}
+			}
+			if (found) {
+				brickList.addAll(idx, bricksToAdd);
+				return true;
+			}
+		}
+
+		for (Brick b : brickList) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java:284]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/Script.java [hunk 10]
```diff
@@ -219,4 +219,73 @@ public void deselectElements(List<UserData<?>> elements) {
...
+
+		if (subStackIdx == -1) {
+			if (scriptId.equals(parentId) || getScriptBrick().getBrickID().equals(parentId)) {
+				brickList.addAll(0, bricksToAdd);
+				return true;
+			}
+
+			boolean found = false;
+			int idx = 0;
+			for (Brick b : brickList) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java:271]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/Script.java [hunk 11]
```diff
@@ -219,4 +219,73 @@ public void deselectElements(List<UserData<?>> elements) {
...
+		// only if the sub stack index is -1, the brick can be a direct child of the script brick
+
+		if (subStackIdx == -1) {
+			if (scriptId.equals(parentId) || getScriptBrick().getBrickID().equals(parentId)) {
+				brickList.addAll(0, bricksToAdd);
+				return true;
+			}
+
+			boolean found = false;
+			int idx = 0;
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java:270]: variable name

File: catroid/src/main/java/org/catrobat/catroid/content/Script.java [hunk 12]
```diff
@@ -219,4 +219,73 @@ public void deselectElements(List<UserData<?>> elements) {
...
+				removeBrick(b);
+			}
+		}
+
+		return bricks;
+	}
+
+	public boolean insertBrickAfter(UUID parentId, int subStackIdx, List<Brick> bricksToAdd) {
+
+		// only if the sub stack index is -1, the brick can be a direct child of the script brick
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java:261]: no comments pls

File: catroid/src/main/java/org/catrobat/catroid/content/Script.java [hunk 13]
```diff
@@ -219,4 +219,73 @@ public void deselectElements(List<UserData<?>> elements) {
...
+			return bricks;
+		}
+		return null;
+	}
+
+	public List<Brick> removeBricksFromScript(List<UUID> brickIds) {
+		List<Brick> bricks = findBricksInScript(brickIds);
+
+		if (bricks != null) {
+			for (Brick b : bricks) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java:251]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/Script.java [hunk 14]
```diff
@@ -219,4 +219,73 @@ public void deselectElements(List<UserData<?>> elements) {
 			}
 		}
 	}
+
+	private List<Brick> findBricksInScript(List<UUID> brickIds) {
+		List<Brick> bricks = new ArrayList<>();
+
+		for (Brick b : brickList) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/Script.java:226]: variable name - no abbreviations please

File: catroid/src/main/java/org/catrobat/catroid/content/DummyScript.kt [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        if (scriptBrick == null) {
+            scriptBrick = DummyBrick(this)
+        }
+        return scriptBrick
+    }
+
+    override fun createEventId(sprite: Sprite?): EventId = EventId(EventId.OTHER)
+
+    override fun run(sprite: Sprite?, sequence: ScriptSequenceAction?) {
+        // do nothing
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/content/DummyScript.kt:43]: pls delete

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/AddBrickCatblocksTest.java [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+		synchronized (addBricksAction) {
+			try {
+				baseActivityTestRule.getActivity().runOnUiThread(addBricksAction);
+				addBricksAction.wait();
+			} catch (InterruptedException e) {
+				e.getMessage(); // ignore
+			}
+			String strId = addedScript.getScriptId().toString().toLowerCase();
+			int tryCounter = 0;
+			// it may take some time until the brick appears in Catblocks, so try multiple times
```
Comment (juliajulie95) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/AddBrickCatblocksTest.java:96]: pls delete

**Review by juliajulie95 — CHANGES_REQUESTED** [review]
https://jenkins.catrob.at/job/Catroid/job/PR-3924/9/analysis/
One string is unused, please delete it

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/AddBrickCatblocksTest.java [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+	public FragmentActivityTestRule<SpriteActivity> baseActivityTestRule = new
+			FragmentActivityTestRule<>(SpriteActivity.class, SpriteActivity.EXTRA_FRAGMENT_POSITION,
+			SpriteActivity.FRAGMENT_SCRIPTS);
+	private WhenScript addedScript = new WhenScript();
+	private Runnable addBricksAction = new Runnable() {
+		@Override
+		public void run() {
+			synchronized (this) {
+				View v = baseActivityTestRule.getActivity().findViewById(R.id.catblocksWebView);
+				Fragment f = FragmentManager.findFragment(v);
```
Comment (juliajulie95) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/AddBrickCatblocksTest.java:63]: abbreviations

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CatblocksScriptFragment.kt [hunk 18]
```diff
@@ -245,5 +251,265 @@ class CatblocksScriptFragment(
...
+            }
+        }
+
+        inner class BrickHelper {
+            var brickIsScript: Boolean = false
+            var script: Script? = null
+            var brick: Brick? = null
+        }
+
+        private fun locateBrickInSprite(sprite: Sprite, brickId: UUID): BrickHelper? {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/CatblocksScriptFragment.kt:339]: duplicate code?

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ScriptFragment.java [hunk 19]
```diff
@@ -816,4 +825,39 @@ public void showUndo(boolean visible) {
...
+		if (getActivity() != null) {
+			int finalScrollToIndex = scrollToIndex;
+			getActivity().runOnUiThread(new Runnable() {
+				@Override
+				public void run() {
+					listView.setSelection(finalScrollToIndex);
+				}
+			});
+		}
+		// ensure scrolling only once
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ScriptFragment.java:859]: comment

**Review by juliajulie95 — CHANGES_REQUESTED** [review]

**Review by juliajulie95 — APPROVED** [review]

**Comment by wslany** [comment]
Thanks!

**Pull request merged**